keps/sig-instrumentation/4200-healthcheck-metrics/README.md

```diff
@@ -0,0 +1,297 @@
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP intends to allow us to emit health check data in a structured and consistent way,
+so that monitoring agents can consume healthz/livez/readyz data and create SLOs/alerts off
+of these inputs.
+
+## Motivation
+
+Healthchecking data is currently surfaced in unstructured format and is scraped by monitoring
+agents (as well as the kubelet), which must be configured to interpret the health data and
+to act upon it. This process does not lend itself readily to the creation of availability SLOs,
+since we basically require an outside agent to parse the health endpoint and convert this signal
+into an SLI. 
+
+### Goals
+
+- Create a uniform interface by which we can consume health checking information
+- Allow availability SLIs to be created without a specialized monitoring agent
+
+
+### Non-Goals
+
+- Creation of SLOs are out of scope. This KEP specifically targets the creation of signals which can be used as SLIs.
+
+## Proposal
+
+We are proposing to add a new endpoint on Kubernetes components `/health/metrics` which returns
+health check data in prometheus format.
```

Comment:
Is there an argument against adding these metrics to the default collector?  Generally being able to grab metrics is equivalent to the default health check, so normally I would say “no”, but if the metrics are useful every 5s they are useful every 15s or 30s.

Reply:
Yeah, health data is basically really susceptible to staleness and given most monitoring configurations scrape every 30s/1m, you'd end up prone to missing edges in health data (i.e. healthz goes unhealthy then blips back or something). We can counter this a bit with using counters (we actually have both counters and gauges for the healthcheck data) but the counter ratios can be thrown off by manual hits to the endpoint, or by processes which repeatedly poll the health endpoint when it is unhealthy (waiting for the process to become healthy). The sole purpose of the dedicated endpoint is to allow for greater granularity of the scrape interval and to allow exposing the actual state of the endpoint at scrape time.

Reply:
I would agree with @logicalhan that splitting metrics endpoints is a correct approach that allows for more granularity on the scrape interval. Ideally, we would want to do that whenever we know that certain metrics require more active scraping than others. For instance, we've already done that in the kubelet with the cpu and memory metrics in order to improve the reactivity autoscaling.

> Is there an argument against adding these metrics to the default collector?

I don't know about the other monitoring backends out there, but Prometheus will treat similar metrics exposed by different endpoints differently by injecting a label containing the path on which the timeserie was scraped. So you will end up deduplicating all the health metrics that you will gather if you add them to the default collector. That's not a big issue considering that the number of new metrics that will be added is very limited, but I don't think that it would be great to deduplicate everything either. I'd rather have very distinct endpoints with good documentation around them.